### PR TITLE
Refactor System.IO.Pipes with mini-PAL

### PR DESCRIPTION
**Pull request by stephentoub**
This separates out the Win32 code used by System.IO.Pipes into its own files; these files are then conditionally included based on the OS property supplied to MSBuild.  Files for Unix are included as well, with stubbed out implementations such that "msbuild /p:OS=Unix" will compile, but it's not yet functional.

Little implementation was actually changed, just moved around. But we should ideally wait to commit this until the Pipes tests have been committed so that these changes can be validated against those.

### CODE REVIEW

**Comment by stephentoub** [comment]
@ellismg

File: src/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Windows.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    {
+        protected override bool ReleaseHandle()
+        {
+            return Interop.mincore.CloseHandle(handle);
+        }
+
+        public override bool IsInvalid
+        {
+            [SecurityCritical]
+            get { return handle == new IntPtr(0) || handle == new IntPtr(-1); }
```
Comment (ellismg) [inline src/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Windows.cs]: NIT: Do you want to use IntPtr.Zero here instead?

Comment (stephentoub) [inline src/System.IO.Pipes/src/Microsoft/Win32/SafeHandles/SafePipeHandle.Windows.cs]: Sure, good suggestion.  This is just the code that was already there, but I'll change it to use IntPtr.Zero instead.

**Comment by ellismg** [comment]
Looks good.

**Comment by mellinoe** [comment]
Looks good to me as well.

**Comment by stephentoub** [comment]
Thanks, @ellismg and @mellinoe.  @mellinoe, do you expect #308 to merge soon?

**Comment by mellinoe** [comment]
There was some talk of moving to more powerful VM's very soon. I would imagine there would be no blockers at that point for merging it in. Keep in mind there are quite a few more tests that have yet to be ported from our internal sources, because they launch multiple processes, something we don't have support for yet with our current testing infrastructure. However, I think this change here should be fine once validated with the "basic tests" that are in #308.

**Comment by stephentoub** [comment]
I merged in the conflicts from master and verified locally against the basic tests from #308, so I'll go ahead and commit this now.

**Pull request merged**